File: java/Detectify.java
package com.detectify;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

public class Detectify {
    /**
     * Detectify API endpoint, no trailing slash
     */
    private static final String Endpoint = "https://api.detectify.com/rest/v2";

    private String apiKey;
    private String secretKey;

    Detectify(String apiKey, String secretKey) {
        this.apiKey = apiKey;
        this.secretKey = secretKey;
    }

    /**
     * Create the HTTP headers for API requests.
     *
     * @param method    The HTTP method to use for the request, in uppercase
     * @param path      The path of the request
     * @param timestamp The timestamp of the request
     * @param body      The optional body of the request
     */
    private Map<String, String> MakeHeaders(String method, String path, Date timestamp, String body) {
        // Signature timestamp uses Unix epoch time
        Long epoch = timestamp.getTime() / 1000;

        // Format hash payload
        String message = String.format("%s;%s;%s;%s;%s", method, path, this.apiKey, epoch, body);

        // Decode base64 secret key to binary
        byte[] key = Base64.getDecoder().decode(this.secretKey);

        // Create the signature
        String signature = "";
        try {
            Mac hasher = Mac.getInstance("HmacSHA256");
            hasher.init(new SecretKeySpec(key, "HmacSHA256"));

            byte[] hash = hasher.doFinal(message.getBytes());

            // Encode signature back to base64
            signature = Base64.getEncoder().encodeToString(hash);
        } catch (NoSuchAlgorithmException e) {
        } catch (InvalidKeyException e) {
        }

        Map<String, String> headers = new HashMap<>();
        headers.put("X-Detectify-Key", this.apiKey);
        headers.put("X-Detectify-Signature", signature);
        headers.put("X-Detectify-Timestamp", epoch.toString());

        return headers;
    }

    public boolean StartScan(String scanProfile) {
        Date timestamp = new Date();

        // Format API request URL
        String path = String.format("/scans/%s/", scanProfile);
        String method = "POST";

        URL url;
        try {
            url = new URL(String.format("%s%s", Detectify.Endpoint, path));
        } catch (IOException e) {
        }

        // Create Detectify signature HTTP headers
        Map<String, String> headers = MakeHeaders(method, path, timestamp, "");

        int statusCode;

        // Call the Detectify API
        try {
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod(method);

            // Add HTTP headers to request
            for (Map.Entry<String, String> header : headers.entrySet()) {
                conn.setRequestProperty(header.getKey(), header.getValue());
            }

            // Get result of request
            statusCode = conn.getResponseCode();

        } catch (IOException e) {
        }

        switch (statusCode) {
            case 202:
                System.out.println("Scan start request accepted");
                return true;
            case 400:
                System.out.println("Invalid scan profile token");
                return false;
            case 401:
                System.out.println("Missing/invalid API key or message signature, or invalid timestamp");
                return false;
            case 403:
                System.out.println("The API key cannot access this functionality");
                return false;
            case 409:
                System.out.println("A scan is already running on the specified profile");
                return false;
            case 423:
                System.out.println("The domain is not verified");
                return false;
            case 500:
                System.out.println("An error occurred while processing the request");
                return false;
            case 503:
                System.out.println("An error occurred while processing the request");
                return false;
            default:
                System.out.println(String.format("Unhandled API response code: %d", statusCode));
                return false;
        }
    }

    public void ScanStatus(String scanProfile) {
        Date timestamp = new Date();

        // Format API request URL
        String path = String.format("/scans/%s/", scanProfile);
        String method = "GET";

        URL url;
        try {
            url = new URL(String.format("%s%s", Detectify.Endpoint, path));
        } catch (IOException e) {
        }

        // Create Detectify signature HTTP headers
        Map<String, String> headers = MakeHeaders(method, path, timestamp, "");

        int statusCode;

        // Buffer for JSON response
        StringBuffer response = new StringBuffer();

        // Call the Detectify API
        try {
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod(method);

            for (Map.Entry<String, String> header : headers.entrySet()) {
                conn.setRequestProperty(header.getKey(), header.getValue());
            }

            // Get result of request
            statusCode = conn.getResponseCode();

            // Read JSON response
            BufferedReader in = new BufferedReader(
                    new InputStreamReader(conn.getInputStream()));
            String inputLine;

            while ((inputLine = in.readLine()) != null) {
                response.append(inputLine);
            }
            in.close();
            conn.disconnect();
        } catch (IOException e) {
        }

        switch (statusCode) {
            case 200:
                System.out.println(response);
                break;
            case 400:
                System.out.println("Invalid scan profile token");
                break;
            case 401:
                System.out.println("Missing/invalid API key or message signature, or invalid timestamp");
                break;
            case 403:
                System.out.println("The API key cannot access this functionality");
                break;
            case 404:
                System.out.println("No scan running for the specified profile, or the specified scan profile does not exist or the API key cannot access the scan profile");
            case 500:
                System.out.println("An error occurred while processing the request");
                break;
            case 503:
                System.out.println("An error occurred while processing the request");
                break;
            default:
                System.out.println(String.format("Unhandled API response code: %d", statusCode));
        }
    }

}

File: c-sharp/detectify.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Threading.Tasks;

namespace DetectifyExample
{
    internal class Program
    {
        public static async Task Main(string[] args)
        {
            // Detectify keys
            var apiKey = "d4bf676ee6146557cbf0f28fe6cbc290";
            var secretKey = "SGVsbG8sIHdvcmxkISBJIGFtIGEgdGVhcG90IQ==";

            // Token for the scan profile
            var scanProfile = "5605b488634efe810dff4276e28ca7f9";

            // Create the API client
            var detectify = new Detectify(apiKey, secretKey);

            // Start a scan
            bool started = await detectify.StartScanAsync(scanProfile);
            if (started)
            {
                // Show scan status if scan successfully started
                await detectify.ScanStatusAsync(scanProfile);
            }
        }
    }

    public class Detectify
    {
        /// <summary>
        /// Detectify API endpoint, without a trailing slash.
        /// </summary>
        private const string Endpoint = "https://api.detectify.com/rest/v2";

        private string ApiKey { get; }
        private string SecretKey { get; }

        public Detectify(string apiKey, string secretKey)
        {
            ApiKey = apiKey;
            SecretKey = secretKey;
        }


        /// <summary>
        /// Create the headers used to sign an API request.
        /// </summary>
        /// <param name="method">The method used for the call, in uppercase.</param>
        /// <param name="path">The path of the request, ie `/v2/domains/`.</param>
        /// <param name="timestamp">The timestamp used when creating the signature.</param>
        /// <param name="body">The body used for requests that require a provided payload. Must be null or an empty string if the request has no body.</param>
        /// <returns>Returns a dictionary of signature headers to use with an API call.</returns>
        private Dictionary<string, string> MakeHeaders(string method, string path,
            DateTime timestamp, string body)
        {
            // Signature timestamp is in Unix epoch format
            var epoch = (long) (timestamp - new DateTime(1970, 1, 1, 0, 0, 0, 0)).TotalSeconds;

            // Calculate signature hash
            var signatureValue = $"{method};{path};{ApiKey};{epoch};{body}";
            var signatureBytes = new HMACSHA256(Convert.FromBase64String(SecretKey))
                .ComputeHash(Encoding.Default.GetBytes(signatureValue));
            var signature = Convert.ToBase64String(signatureBytes);

            return new Dictionary<string, string>
            {
                ["X-Detectify-Key"] = ApiKey,
                ["X-Detectify-Signature"] = signature,
                ["X-Detectify-Timestamp"] = epoch.ToString()
            };
        }

        /// <summary>
        /// Start a scan for a given scan profile.
        /// </summary>
        /// <param name="scanProfile">The scan profile to start a scan on.</param>
        /// <returns>Returns true if a scan was started, false if not.</returns>
        public async Task<bool> StartScanAsync(string scanProfile)
        {
            var path = $"/scans/{scanProfile}/";
            var url = $"{Endpoint}{path}";
            var timestamp = DateTime.UtcNow;

            // Create Detectify headers
            var headers = MakeHeaders("POST", path, timestamp, null);

            using (var client = new HttpClient())
            {
                // Add Detectify headers to request
                headers.ToList().ForEach(h => client.DefaultRequestHeaders.Add(h.Key, h.Value));

                var response = await client.PostAsync(url, null);

                switch ((int) response.StatusCode)
                {
                    case 202:
                        Console.WriteLine("Scan start request accepted");
                        return true;
                    case 400:
                        Console.WriteLine("Invalid scan profile token");
                        return false;
                    case 401:
                        Console.WriteLine("Missing/invalid API key or message signature, or invalid timestamp");
                        return false;
                    case 403:
                        Console.WriteLine("The API key cannot access this functionality");
                        return false;
                    case 404:
                        Console.WriteLine(
                            "The specified scan profile does not exist or the API cannot access the profile");
                        return false;
                    case 409:
                        Console.WriteLine("A scan is already running on the specified profile");
                        return false;
                    case 423:
                        Console.WriteLine("The domain is not verified");
                        return false;
                    case 500:
                    case 503:
                        Console.WriteLine("An error occurred while processing the request");
                        return false;
                    default:
                        Console.WriteLine($"API returned unhandled status code: {(int) response.StatusCode}");
                        return false;
                }
            }
        }

        /// <summary>
        /// Retrieves the status of a currently running scan for a given scan profile.
        /// </summary>
        /// <param name="scanProfile">The scan profile token to check scan status for.</param>
        public async Task ScanStatusAsync(string scanProfile)
        {
            var path = $"/scans/{scanProfile}/";
            var url = $"{Endpoint}{path}";
            var timestamp = DateTime.UtcNow;
            
            // Create Detectify headers
            var headers = MakeHeaders("GET", path, timestamp, null);

            using (var client = new HttpClient())
            {
                // Add Detectify headers to request
                headers.ToList().ForEach(h => client.DefaultRequestHeaders.Add(h.Key, h.Value));

                var response = await client.GetAsync(url);

                switch ((int) response.StatusCode)
                {
                    case 200:
                        Console.WriteLine(await response.Content.ReadAsStringAsync());
                        break;
                    case 400:
                        Console.WriteLine("Invalid scan profile token");
                        break;
                    case 401:
                        Console.WriteLine("Missing/invalid API key or message signature, or invalid timestamp");
                        break;
                    case 403:
                        Console.WriteLine("The API key cannot access this functionality");
                        break;
                    case 404:
                        Console.WriteLine(
                            "No scan running for the specified profile, or the specified scan profile does not exist, or the API cannot access the profile");
                        break;
                    case 500:
                    case 503:
                        Console.WriteLine("An error occurred while processing the request");
                        break;
                    default:
                        Console.WriteLine($"API returned unhandled status code: {(int) response.StatusCode}");
                        break;
                }
            }
        }
    }
}
